### ISSUE TITLE
How to preserve Homekit Scenes, Automations, Rooms and Associations for knx accessories

### ISSUE THREAD

**Issue by migabc** [snowdd1/homebridge-knx#183]
Is there an easy way to preserve my Homekit configuration (scenes, automations, rooms, associations, etc) for my knx accessories?

Every time I try to upgrade this plugin and something goes wrong (for whatever reason) even if I later rollback the plugin, along with all the knx_config.json, persist and accessories files I still end up losing all my Homekit scenes, automations, rooms and associations related to these knx devices

A a simple backup and restore of files (or even the whole image) will not preserve my Homekit configs 

This quite frustrating as I have spent several hours tweeking all my Homekit configs :-(

This problem only only happens with this homebridge-knx plugin
This does not happen with my other homebridge plugins

**Comment by snowdd1** [comment]
It should. The `knx_config.json` file gets modified during runtime (first start), so you need to backup that one **after** the first start, otherwise all devices get new IDs after the restore.

**Comment by migabc** [comment]
Nope. It doesn't.
Please see what I wrote above:

ven if I later rollback the plugin, along with all the knx_config.json, persist and accessories files I still end up losing all my Homekit scenes, automations, rooms and associations related to these knx devices

A a simple backup and restore of files (or even restoring the whole image) will not restore my lost/removed Homekit configs caused by homebridge-knx failed to start up properly

**Comment by snowdd1** [comment]
That is strange though. I just moved my whole homebridge instance to a raspberry without loosing anything, even with a new IP address and hardware address on the new hardware. The homebridge instance was immediately recognized by HomeKit.

**Comment by migabc** [comment]
Yes, that's true.
But if for some reason the knx plugin does not get registered when homebridge starts up, all the knx accessories, scenes, automations, etc will be completely wiped out from Homekit

Even if you rollback (recover your initial config), you will only recover the knx accessories themselves (which will now stay in the default room) - all your previous configurations (Rooms, scenes, automations, etc) related to those accessories will be completely lost

Is there some way of backing up/restoring your Homekit config?

**Comment by snowdd1** [comment]
That is right and can't be changed AFAIK. Because that is HomeKit clearing its database in the iCloud.

**Comment by migabc** [comment]
This problem happend to me when I tried to upgrade homebridge to the lastest version 1.0.6 (needed for someother pluin) - However, this latest version of homebridge (in fact any version  >= 1.0) does not support the knx plugin - so it just started up without registering the knx plugin

So, boom - all my 200+ knx accessories were lost 
Damn :-(

**Comment by migabc** [comment]
P.S: You should put a note stating not to upgrade to latest homebridge version (or at least the necessary configs updates before doing so)

**Comment by migabc** [comment]
Anyway, so there's no way of backing up/restoring the HomeKit database in the iCloud?
It would be good if we could do some sort of TimeMachine to get things back to the way they were, say "Yessterday"
I'm sure I'm not the only one having this issue

**Comment by giase82** [comment]
I am using the “Controller for Home” App, which allows to backup and restore to some extent. It is not flawless but works for most of the things and can safe some work when restoring...
https://apps.apple.com/de/app/controller-for-homekit/id1198176727?l=en

//Christof

**Comment by fred925** [comment]
ref. migabc's comment Nov 12, 2020
_"However, this latest version of homebridge (in fact any version >= 1.0) does not support the knx plugin "_

I'm using homebridge v. 1.0.7 and have been struggling with the knx plugin homebridge-knx-easy@0.3.40. 
I can see in the homebridge log that the plugin is loaded, however I can not see any reference to the knx_config.json containing a simple device; "ServiceType": "Lightbulb". 
No accessories appear in homebridge, also when inspecting the knx_config.json after "first start" i see no UID added.
(knxd is working; I can switch on/off the "Lightbulb"from command line w. knxtool)

What is the latest homebridge version to support knx plugin, and which knx plugin should be used?

**Comment by snowdd1** [comment]
@fred925 :

Sorry I cannot support any forks like  homebridge-knx-easy.
You might:
* talk to @mariomaurer to create pull requests with his improvements
* revert to the original

![image](https://user-images.githubusercontent.com/11786396/104159119-0a4c6500-53ef-11eb-8762-abdd5223504e.png)

**Comment by migabc** [comment]
This happens because if homebridge starts without registering the homebridge-knx plugin it will remove all knx cached accessories from the cachedAccessories file

I found a workaorund.
To prevent this you need to remove the write access on the root user for this file:

```
sudo find / -name cachedAccessories
sudo chmod 444 /var/lib/homebridge/accessories/cachedAccessories
```

So now whenever an upgrade (example nodejs, etc) is done on the system that prevents homebridge from registering homebridge-knx, the homebridge when starting up will not remove the knx accessories from cachedAccessories and this way we have enough time to troubleshoot the issue and eventually rollback that upgrade that prevented the homebridge-knx plugin from registering without losing all our knx accessories

Note:
Whenever you need to reconfigure/update either the config.json or knx_config.json you must re-enable write access on that file by:

`
sudo chmod 644 /var/lib/homebridge/accessories/cachedAccessories
`

After that the config.json or knx_config.json will get modified during runtime (only during the first start after the config change), so after that first start after you modified your config, you can then remove the write access on the cachedAccessories file again
